File: strategy/typescript/glados/task/attacker/midfield.ts
import { Position, Vector } from "base/vector";

import { MessageType } from "glados/control/messaging";
import * as Physics from "glados/observer/physics";
import { MidfieldSampling } from "glados/task/ability/midfieldsampling";
import { SuggestPass } from "glados/task/ability/suggestpass";
import { Agent, Task } from "glados/task/base";
import * as PathHelper from "glados/trajectory/pathhelper";
import { ToTarget } from "glados/trajectory/totarget";


export class Midfield extends Task {
	private _passPos: Position | undefined = undefined; // ewwwww hack
	private _frameCount: number = 0;
	private _obstacleTable: PathHelper.PathHelperParameters;

	private _suggestPass: SuggestPass;
	private _midfieldSampling: MidfieldSampling;

	constructor(agent: Agent) {
		super(agent);

		let ignore = false;
		this._obstacleTable = {
			ignoreBall: ignore,
			ignoreGoals: ignore,
			ignoreDefenseArea: ignore,
			ignoreOpponentDefenseArea: ignore,
			messaging: this._messaging,
			ignorePass: ignore,
			ignoreBallPlacementObstacle: false
		};

		this._suggestPass = new SuggestPass(this._robot, this._messaging);
		this._midfieldSampling = new MidfieldSampling(this._robot, this._messaging);
	}

	private _samplePassPosition(): Position {
		let zone = this._messaging.receiveTrainer(MessageType.midfieldZone);
		if (zone == undefined) {
			throw new Error("midfield task running without zone assignment");
		}

		let left = zone.boundaries.left;
		let right = zone.boundaries.right;
		let top = zone.boundaries.top;
		let bottom = zone.boundaries.bottom;

		let width = right - left;
		let height = top - bottom;

		let xStep = width / 3;
		let yStep = height / 6;

		let bestScore = -Infinity;
		let bestPoint = undefined;
		for (let x = left; width < 0 ? x >= left + width : x <= left + width;x += xStep) {
			for (let y = bottom; height < 0 ? y >= bottom + height : y <= bottom + height; y += yStep) {
				let candidatePoint = new Vector(x, y);
				let rating = this._midfieldSampling.evalLocation(candidatePoint, bestScore);
				if (rating > bestScore) {
					bestScore = rating;
					bestPoint = candidatePoint;
				}
			}
		}

		return bestPoint!;
	}

	// local disco = [
	// 	vis.colors.red,
	// 	vis.colors.blue,
	// 	vis.colors.green,
	// 	vis.colors.pink,
	// 	vis.colors.turquoise,
	// 	vis.colors.yellow,
	// 	vis.colors.skyBlue,
	// 	vis.colors.mediumPurple
	// ]

	run() {
		PathHelper.setDefaultObstaclesByTable(this._robot.path, this._robot, this._obstacleTable);

		this._midfieldSampling.precalculate();

		// Hacky quickfix for messaging delay problems
		if ((this._frameCount % 2) === 0) {
			this._passPos = this._samplePassPosition();
		}
		this._frameCount = this._frameCount + 1;

		// local random = Math.round(Math.random() * #disco)
		// vis.addCircle("middy", this._robot.pos, 0.1, disco[random] or vis.colors.orange, true)

		let zone = this._messaging.receiveTrainer(MessageType.midfieldZone);
		let defaultPos = zone!.defaultPos;

		let attackPosition = this._messaging.receiveSingleSender(MessageType.attackPosition)[1];

		let time = Physics.robotTimeToPos(this._robot, this._passPos!, new Vector(0, 0))[0];
		if (this._passPos) {
			this._suggestPass._suggestPass(this._passPos, attackPosition, time);
		}

		this._robot.trajectory.update(ToTarget, defaultPos, Math.PI / 2, undefined, new Vector(0, 0));
	}
}

File: strategy/lua/marvin/task/attacker/midfield.lua
--[[***********************************************************************
*   Copyright 2026 Robotics Erlangen e.V.                                 *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local SuggestPass = require "task/ability/suggestpass"
local MidfieldSampling = require "task/ability/midfieldsampling"
local Midfield = Class("Task.Midfield", require "task/base", SuggestPass, MidfieldSampling)

local Physics = require "observer/physics"
local PathHelper = require "trajectory/pathhelper"
local ToTarget = require "trajectory/totarget"

function Midfield:_init()
	self._passPos = nil

	-- ewwwww hack
	self._frameCount = 0

	local ignore = false
	self._obstacleTable = {
		ignoreBall = ignore,
		ignoreGoals = ignore,
		ignoreDefenseArea = ignore,
		ignoreOpponentDefenseArea = ignore,
		inbox = self._inbox,
		ignorePass = (not self._inbox) or ignore,
		ignoreBallPlacementObstacle = false
	}
end

function Midfield:_samplePassPosition()
	local zone = self._inbox.midfieldZone().trainer

	local left = zone.boundaries.left
	local right = zone.boundaries.right
	local top = zone.boundaries.top
	local bottom = zone.boundaries.bottom

	local width = right - left
	local height = top - bottom

	local xStep = width / 3
	local yStep = height / 6

	local bestScore = -math.huge
	local bestPoint = nil
	for x = left, left + width, xStep do
		for y = bottom, bottom + height, yStep do
			local candidatePoint = Vector(x, y)
			local rating = self:evalLocation(candidatePoint, bestScore)
			if rating > bestScore then
				bestScore = rating
				bestPoint = candidatePoint
			end
		end
	end

	return bestPoint
end

-- local disco = {
-- 	vis.colors.red,
-- 	vis.colors.blue,
-- 	vis.colors.green,
-- 	vis.colors.pink,
-- 	vis.colors.turquoise,
-- 	vis.colors.yellow,
-- 	vis.colors.skyBlue,
-- 	vis.colors.mediumPurple
-- }

function Midfield:run()
	PathHelper.setDefaultObstaclesByTable(self._robot.path, self._robot, self._obstacleTable)

	self:precalculate()

	-- Hacky quickfix for messaging delay problems
	if (self._frameCount % 2) == 0 then
		self._passPos = self:_samplePassPosition()
	end
	self._frameCount = self._frameCount + 1

	-- local random = math.round(math.random() * #disco)
	-- vis.addCircle("middy", self._robot.pos, 0.1, disco[random] or vis.colors.orange, true)

	local zone = self._inbox.midfieldZone().trainer
	local defaultPos = zone.defaultPos

	local _, attackPosition = next(self._inbox.attackPosition())

	local time = Physics.robotTimeToPos(self._robot, self._passPos, Vector(0, 0))
	if self._passPos then
		self:_suggestPass(self._passPos, attackPosition, time)
	end
	
	self._robot.trajectory:update(ToTarget, defaultPos, math.pi/2, nil, Vector(0, 0))
end


return Midfield
